---
title: "The Human Lung Cell Atlas (10X)"
output:
  html_document: default
  html_notebook: default
---
# All Package versions used


```{r, fig.width=10, fig.height=10}
installed.packages()[,'Version']

```

# Specify the tissue of interest, run the boilerplate code which sets up the functions and environment, load the tissue object.


```{r, fig.width=10, fig.height=10}
tissue_of_interest = "all"
library(here)
source(here("boilerplate.R"))
tiss10x <- load_tissue_droplet(tissue_of_interest)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = tiss10x, pc.use = 1:9, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = tiss10x, pc.use = 10:18, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
```

```{r, fig.width=10, fig.height=10}
PCElbowPlot(object = tiss10x)
```

```{r, fig.width=10, fig.height=10}
# Set number of principal components.
n.pcs = 20
```

```{r, fig.width=10, fig.height=10}
# Set resolution
res.used <- 2

tiss10x <- FindClusters(object = tiss10x, reduction.type = "pca", dims.use = 1:n.pcs,
    resolution = res.used, print.output = 0, save.SNN = TRUE)
```

```{r, fig.width=10, fig.height=10}
tiss10x <- RunTSNE(object = tiss10x, dims.use = 1:n.pcs, seed.use = 10, perplexity=30)
```


```{r, fig.width=10, fig.height=10}
# note that you can set do.label=T to help label individual clusters
TSNEPlot(object = tiss10x, do.label = T, no.legend = T)
```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = tiss10x, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = tiss10x, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = tiss10x, do.return = TRUE, group.by = "region")
TSNEPlot(object = tiss10x, do.return = TRUE, group.by = "location")
TSNEPlot(object = tiss10x, do.return = TRUE, group.by = "compartment")
```

Check expression of genes of interset.

```{r, fig.height=10, fig.width=10}
genes_to_check = toupper(c('Cldn5', 'Epcam', 'Col1a2', 'Ptprc'))
FeaturePlot(tiss10x, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)
```

How big are the clusters?
```{r, fig.width=10, fig.height=10}
table(tiss10x@ident)
```

# Subclustering normal cells

```{r, fig.width=10, fig.height=10}
ntiss10x <- SubsetData(tiss10x, cells.use = rownames(tiss10x@meta.data[which(tiss10x@meta.data[,'region'] %in% c('normal')),]))

```

```{r, fig.width=10, fig.height=10}
ntiss10x <- ntiss10x %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30 ) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 30
normal.res.use = 2
ntiss10x <- ntiss10x %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30, force.recalc = TRUE)

TSNEPlot(object = ntiss10x, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=10, fig.width=10}

genes_to_check = toupper(c('Cldn5', 'Epcam', 'Col1a2', 'Ptprc'))

FeaturePlot(ntiss10x, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

# Subclustering normal stromal cells

```{r, fig.width=10, fig.height=10}
ntiss10x.stromal = SubsetData(ntiss10x, ident.use = c(27,17,21))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.stromal <- ntiss10x.stromal %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.stromal, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.stromal, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCElbowPlot(ntiss10x.stromal)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 10
normal.res.use = 2.5
ntiss10x.stromal <- ntiss10x.stromal %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)

TSNEPlot(object = ntiss10x.stromal, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.stromal, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.stromal, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.stromal, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.stromal, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.stromal, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=50, fig.width=10}

genes_to_check = toupper(c('Cldn5', 'Epcam', 'Col1a2', 'Ptprc', 'Serpinf1', 'Plin2', 'Apoe', 'PDGFRA', 'PDGFRB', 'CNN1', 'TRPC6', 'SLC38a5', 'PI16', 'Serpinf1', 'HP', 'FGF18', 'SCARA5', 'MSLN', 'WIF1', 'Des', 'KCNA5', 'ADAMTS4', 'Acta2', 'Angptl4'))

FeaturePlot(ntiss10x.stromal, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.stromal.markers <- FindAllMarkers(ntiss10x.stromal, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=80, fig.width=10}
genes_to_check <- ntiss10x.stromal.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.stromal.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.stromal, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}
#At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
#Relying on lessons from mouse to call populations (a few representative markers are shown). All stromal cells are CD45- #CD31- EpCAM-
#COX4I2+ HIGD1B+ GJA4+ TAGLN- CNN- -> Pericyte
#SERPINF1+ PI16+ FBLN1+ SCARA5+ SLC38A5- -> Adventitial Fibroblast
#ACTA2+ TAGLN+ CNN1+ MYH11+ DES+ KCNA5+ -> Airway Smooth Muscle
#SLC38A5+ GDF10+ GPC3+ ITGA8+ -> Alveolar Fibroblast
#PLIN2+ APOE+ -> Lipofibroblast
#MSLN+ KRT19+ UPK3B+ -> Mesothelial
#ACTA2+ MYH11+ ASPN+ TYRP1+ -> Myofibroblast


# stash current cluster IDs
ntiss10x.stromal <- StashIdent(object = ntiss10x.stromal, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.stromal@ident)))
free_annotation <- c("Pericyte",
                     "Adventitial Fibroblast",
                     "Airway Smooth Muscle",
                     "Adventitial Fibroblast",
                     "Airway Smooth Muscle",
                     "Alveolar Fibroblast",
                     "Lipofibroblast",
                     "Myofibroblast",
                     "Mesothelial",
                     "Adventitial Fibroblast")

ntiss10x.stromal@meta.data[,'free_annotation'] <- NA
ntiss10x.stromal@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.stromal@ident, from = cluster.ids, to = free_annotation))

TSNEPlot(ntiss10x.stromal, group.by = 'free_annotation', no.legend = TRUE, do.label = TRUE)
```

# Subclustering normal epithelial cells

```{r, fig.width=10, fig.height=10}
ntiss10x.epi = SubsetData(ntiss10x, ident.use = c(8, 22, 29, 26, 25))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi <- ntiss10x.epi %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.epi, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.epi, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCElbowPlot(ntiss10x.epi)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 15
normal.res.use = 1
ntiss10x.epi <- ntiss10x.epi %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.epi, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.epi, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.epi, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.epi, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.epi, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.epi, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=50, fig.width=10}

genes_to_check = toupper(c('Cldn5', 'Epcam', 'Col1a2', 'Ptprc', 'Muc5ac', 'Muc5b', 'SFTPC', 'ETV5', 'PDPN', 'HOPX', 'CLIC5', 'FOXJ1', 'TUBB4B', 'SCGB1A1', 'SCGB3A2', 'WIF1', 'HHIP', 'SERPINA1', 'KRT5', 'NGFR', 'KRT14'))

FeaturePlot(ntiss10x.epi, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.markers <- FindAllMarkers(ntiss10x.epi, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=60, fig.width=10}
genes_to_check <- ntiss10x.epi.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.epi.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.epi, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters

```{r, fig.width=10, fig.height=10}

# Relying on canonical markers to call populations (a few representative markers are shown). All epithelial cells are CD45- CD31- Col1A1-
# SFP*+ ETV5+ MUC1+ -> Alveolar Epithelial Type 2
# HOPX+ PDPN+ CLIC5+ SPC*- -> Alveolar Epithelial Type 1
# SCGB1A1+ SCGB3A2+ SP*- FOXJ1- -> Club
# FOXJ1+ TUBB4+ -> Ciliated
# KRT5+ KRT14+ -> Basal (clustered again in Round 2)


# stash current cluster IDs
ntiss10x.epi <- StashIdent(object = ntiss10x.epi, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.epi@ident)))
free_annotation <- c("Alveolar Epithelial Type 2",
                     "Club",
                     "Alveolar Epithelial Type 2",
                     "Alveolar Epithelial Type 2",
                     NA,
                     "Ciliated",
                     "Alveolar Epithelial Type 1")

ntiss10x.epi@meta.data[,'free_annotation'] <- NA
ntiss10x.epi@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.epi@ident, from = cluster.ids, to = free_annotation))
```



# Subclustering normal epithelial cells (Round 2, Basal Cells)

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r2 = SubsetData(ntiss10x.epi, ident.use = c(4))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r2 <- ntiss10x.epi.r2 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.epi.r2, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.epi.r2, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCElbowPlot(ntiss10x.epi.r2)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 5
normal.res.use = 1
ntiss10x.epi.r2 <- ntiss10x.epi.r2 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE, plot.SNN = TRUE, prune.SNN = 1/2) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.epi.r2, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.epi.r2, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.epi.r2, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.epi.r2, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.epi.r2, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.epi.r2, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=50, fig.width=10}

genes_to_check = toupper(c('Pecam1', 'Epcam', 'Col1a1', 'Acta2', 'Ptprc', 'Muc5ac', 'Muc5b', 'SFTPC', 'ETV5', 'PDPN', 'CLIC5', 'FOXJ1', 'TUBB4B', 'SCGB1A1', 'WIF1', 'HHIP', 'SERPINA1', 'KRT5', 'NGFR', 'SPRR3', 'Hes1', 'KRT14', 'Krt13', 'MKI67'))

FeaturePlot(ntiss10x.epi.r2, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r2.markers <- FindAllMarkers(ntiss10x.epi.r2, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=30, fig.width=10}
genes_to_check <- ntiss10x.epi.r2.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.epi.r2.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.epi.r2, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}

# There is still considerable heterogeneity among the cells, but we don't have enough cells to go further with Seurat
# KRT5+ KRT14+ -> Basal
# KRT7+ KRT13+ -> Differentiating Basal
# MKI67+ -> Proliferating Basal

# stash current cluster IDs
ntiss10x.epi.r2 <- StashIdent(object = ntiss10x.epi.r2, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.epi.r2@ident)))
free_annotation <- c("Differentiating Basal",
                     "Differentiating Basal",
                     "Basal",
                     "Proliferating Basal")

ntiss10x.epi.r2@meta.data[,'free_annotation'] <- NA
ntiss10x.epi.r2@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.epi.r2@ident, from = cluster.ids, to = free_annotation))
```


# Subclustering normal endothelial cells

```{r, fig.width=10, fig.height=10}
cells.to.use = WhichCells(ntiss10x, ident = c(11,5,23,15,14,28))
cells.to.use = intersect(intersect(cells.to.use, names(which(ntiss10x@data['PTPRC',cells.to.use] < 0.1))), names(which(ntiss10x@data['EPCAM',cells.to.use] < 0.1)))


ntiss10x.endo = SubsetData(ntiss10x, cells.use = cells.to.use)
```


```{r, fig.width=10, fig.height=10}
ntiss10x.endo <- ntiss10x.endo %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.endo, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.endo, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCElbowPlot(ntiss10x.endo)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 11
normal.res.use = 1.25
ntiss10x.endo <- ntiss10x.endo %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.endo, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.endo, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.endo, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.endo, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.endo, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.endo, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=45, fig.width=10}

genes_to_check = toupper(c('Cldn5', 'Epcam', 'Col1a2', 'Ptprc', 'Cxcl12', 'Dkk2', 'Gja5', 'Slc6a4', 'Gpihbp1', 'Aplnr', 'Apln', 'Ednrb', 'Myc', 'Spry1', 'Ackr1', 'Rspo3', 'Ptgis', 'Prox1', 'PDPN', 'CTNNB1', 'FOS'))

FeaturePlot(ntiss10x.endo, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.endo.markers <- FindAllMarkers(ntiss10x.endo, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=80, fig.width=10}
genes_to_check <- ntiss10x.endo.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.endo.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.endo, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}
#
# Relying on lessons from mouse to call populations (a few representative markers are shown). All endothelial cells are CD45- EpCAM- Col1A1-
# IL7R+ SLC6A4+ FCN3+ C7- CPE- ACKR1- -> Capillary 2
# APLN+ EDNRB+ HPGD+ IL1RL1+  -> Capillary 1
# CXCL12+ DKK2+ GJA5+ SERPINE2+ -> Artery
# RSPO+ ACKR1+ PTGIS+ PTDGS+ CPE+ -> Vein
# TFF3+ PDPN+ CCL21+ IGF1+ -> Lymphatics
# 
# Not seen in mouse:
# 
# MYC+ SPRY1+ ACKR1+ -> Bronchial Vessel 1
# MYC+ SPRY1- -> Bronchial Vessel 2


# stash current cluster IDs
ntiss10x.endo <- StashIdent(object = ntiss10x.endo, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.endo@ident)))
free_annotation <- c("Capillary 2",
                     "Capillary 1",
                     "Capillary 2",
                     "Artery",
                     "Vein",
                     "Bronchial Vessel 1",
                     "Bronchial Vessel 2",
                     "Lymphatics")

ntiss10x.endo@meta.data[,'free_annotation'] <- NA
ntiss10x.endo@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.endo@ident, from = cluster.ids, to = free_annotation))
```

# Subclustering normal immune cells

```{r, fig.width=10, fig.height=10}
ntiss10x.immune = SubsetData(ntiss10x, ident.use = c(0,1,2,12,9,6,3,18,16,10,7,4,19,24,20))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune <- ntiss10x.immune %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 24
normal.res.use = 1.5
ntiss10x.immune <- ntiss10x.immune %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)

TSNEPlot(object = ntiss10x.immune, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.immune, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=40, fig.width=10}

genes_to_check = toupper(c('PTPRC', 'PECAM1', 'EPCAM', 'COL1A1', 'CD3E', 'KLRB1', 'CD79A', 'CD19', 'MS4A1', 'CD24', 'CD68', 'FCGR3A', 'CD14', 'MARCO', 'MSR1', 'MRC1', 'MKI67'))


FeaturePlot(ntiss10x.immune, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.markers <- FindAllMarkers(ntiss10x.immune, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=500, fig.width=10}
genes_to_check <- ntiss10x.immune.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}

# Calling immune cells based on canonical markers and bulk RNAseq data from the Weissman Lab. All immune cells are EpCAM- Col1a1-, some are CD31+
# CD79A+ CD24+ MS4A1+ CD19+ -> B


# stash current cluster IDs
ntiss10x.immune <- StashIdent(object = ntiss10x.immune, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune@ident)))
free_annotation <- c(NA,
                     NA,
                     NA,
                     NA,
                     NA,
                     #5
                     NA,
                     NA,
                     NA,
                     NA,
                     NA,
                     #10
                     NA,
                     NA,
                     NA,
                     NA,
                     NA,
                     #15
                     NA,
                     "B",
                     NA,
                     NA,
                     NA,
                     NA)

ntiss10x.immune@meta.data[,'free_annotation'] <- NA
ntiss10x.immune@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune@ident, from = cluster.ids, to = free_annotation))
```

# Subclustering normal immune cells (T cells)

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells = SubsetData(ntiss10x.immune, ident.use = c(2,3,5,7,14))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells <- ntiss10x.immune.Tcells %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.Tcells, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.Tcells, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.Tcells, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.Tcells, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 20
normal.res.use = 1
ntiss10x.immune.Tcells <- ntiss10x.immune.Tcells %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)

TSNEPlot(object = ntiss10x.immune.Tcells, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.Tcells, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.Tcells, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.Tcells, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.Tcells, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.immune.Tcells, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=40, fig.width=10}

genes_to_check = c('CD3E', 'CD4', 'IL7R', 'CD8A', grep('^KLR', rownames(ntiss10x.immune@data), value = T))

FeaturePlot(ntiss10x.immune.Tcells, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells.markers <- FindAllMarkers(ntiss10x.immune.Tcells, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=70, fig.width=10}
genes_to_check <- ntiss10x.immune.Tcells.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.Tcells.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.Tcells, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}

# FCER1G+ TYROBP+ GZMB+ CHST2+ HOPX+ CD3E- -> Natural Killer
# CD3E+ CD4+ LTB+ COTL1+ LDHB+ CCR7- LEF1- -> CD4+ Nonnaive T
# CD3E+ CD4+ LTB+ LDHB+ CCR7+ LEF1+ COTL1- -> CD4+ Naive T
# CD3E+ CD8+ GZMH+ GZMB+ LTB- IL7R- GZMK- DUSP2- -> CD8+ Naive T
# CD3E+ CD8+ GZMK+ KLRB1+ IL7R+ DUSP2+ GZMH- GZMB- -> CD8+ Nonnaive T


# stash current cluster IDs
ntiss10x.immune.Tcells <- StashIdent(object = ntiss10x.immune.Tcells, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune.Tcells@ident)))
free_annotation <- c(NA,
                     "Natural Killer",
                     "CD4+ Nonnaive T",
                     "CD4+ Naive T",
                     "CD8+ Nonnaive T",
                     #5
                     "CD8+ Naive T",
                     "Natural Killer")

ntiss10x.immune.Tcells@meta.data[,'free_annotation'] <- NA
ntiss10x.immune.Tcells@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune.Tcells@ident, from = cluster.ids, to = free_annotation))
```



# Subclustering normal immune cells (T Cells, Round 2)

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells.r2 = SubsetData(ntiss10x.immune.Tcells, ident.use = c(0))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells.r2 <- ntiss10x.immune.Tcells.r2 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.Tcells.r2, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.Tcells.r2, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.Tcells.r2, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.Tcells.r2, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 5
normal.res.use = 1
ntiss10x.immune.Tcells.r2 <- ntiss10x.immune.Tcells.r2 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE, plot.SNN = TRUE, prune.SNN = 1/15) %>%
      RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)


TSNEPlot(ntiss10x.immune.Tcells.r2, label.size = 4, pt.size = 1.2, do.label = T)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.Tcells.r2, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.Tcells.r2, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.Tcells.r2, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.Tcells.r2, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.immune.Tcells.r2, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=50, fig.width=10}

genes_to_check = c('CD3E', 'CD4', 'CD8A', 'GZMK', 'IL7R', 'GZMH', grep('^KLR', rownames(ntiss10x.immune@data), value = T))

FeaturePlot(ntiss10x.immune.Tcells.r2, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells.r2.markers <- FindAllMarkers(ntiss10x.immune.Tcells.r2, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=50, fig.width=10}
genes_to_check <- ntiss10x.immune.Tcells.r2.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.Tcells.r2.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.Tcells.r2, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}

# CD3E+ CD8+ GZMH+ GZMB+ LTB- IL7R- GZMK- DUSP2- -> CD8+ Naive T
# CD3E+ CD8+ GZMK+ KLRB1+ IL7R+ DUSP2+ GZMH- GZMB- -> CD8+ Nonnaive T


# stash current cluster IDs
ntiss10x.immune.Tcells.r2 <- StashIdent(object = ntiss10x.immune.Tcells.r2, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune.Tcells.r2@ident)))
free_annotation <- c("CD8+ Nonnaive T",
                     "CD8+ Naive T",
                     "CD8+ Naive T",
                     "CD8+ Naive T",
                     "CD8+ Naive T")

ntiss10x.immune.Tcells.r2@meta.data[,'free_annotation'] <- NA
ntiss10x.immune.Tcells.r2@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune.Tcells.r2@ident, from = cluster.ids, to = free_annotation))
```


# Subclustering normal immune cells (Round 2, Non-T Cells)

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r2 = SubsetData(ntiss10x.immune, ident.use = c(10,4,0,1,15,18,20,6,8,17,13,19,9,12,11))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r2 <- ntiss10x.immune.r2 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.r2, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r2, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r2, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.r2, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 20
normal.res.use = 1.5
ntiss10x.immune.r2 <- ntiss10x.immune.r2 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)

TSNEPlot(object = ntiss10x.immune.r2, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.r2, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.r2, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.r2, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.r2, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.immune.r2, do.return = TRUE, group.by = "compartment")
```


```{r, fig.height=50, fig.width=10}

genes_to_check = toupper(c('MARCO', 'FCGR3B', 'CD68', 'MSR1', 'MRC1', 'FCGR3A', 'CD14', 'CLEC10A', 'GPR183', 'BASP1', 'F13A1', 'SEPP1', 'RNASE1', 'LTF', 'LCN2', 'HP', 'CD24', 'CD19', 'MS4A1', 'CD3E', 'CD74', 'B2M', 'CX3CR1', 'MKI67', 'TOP2A'))


FeaturePlot(ntiss10x.immune.r2, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r2.markers <- FindAllMarkers(ntiss10x.immune.r2, min.pct = 0.25, thresh.use = 0.25, only.pos = T, max.cells.per.ident = 200)

```


```{r, fig.height=175, fig.width=10}
genes_to_check <- ntiss10x.immune.r2.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.r2.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.r2, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters

```{r, fig.width=10, fig.height=10}

# MARCO+ MRC1+ MSR1+ CD68 CD74+ CD14- FCGR3A- -> Macrophage
# MKI67+ TOP2A+ STM1+ MARCO+ MRC1+ MSR1+ CD68 CD74+ CD14- FCGR3A- -> Proliferating Macrophage

# stash current cluster IDs
ntiss10x.immune.r2 <- StashIdent(object = ntiss10x.immune.r2, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune.r2@ident)))
free_annotation <- c("Macrophage",
                     #0
                     "Macrophage",
                     "Macrophage",
                     NA,
                     "Macrophage",
                     NA, #Throwing these out, look like NK/T/B but with much lower expression
                     #5
                     "Macrophage",
                     "Macrophage",
                     "Macrophage",
                     NA,
                     NA,
                     #10
                     "Macrophage",
                     "Macrophage",
                     NA,
                     "Mast Cell/Basophil 1",
                     NA,
                     #15
                     NA,
                     "Macrophage",
                     "Proliferating Macrophage")
                     #18

ntiss10x.immune.r2@meta.data[,'free_annotation'] <- NA
ntiss10x.immune.r2@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune.r2@ident, from = cluster.ids, to = free_annotation))
```


# Subclustering normal immune cells (Round 3, Macrophages)

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r3 = SubsetData(ntiss10x.immune.r2, ident.use = c(4,7,17,12,11,0,8,2,6,1))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r3 <- ntiss10x.immune.r3 %>% ScaleData(vars.to.regress = 'channel') %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.r3, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r3, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r3, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.r3, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 7
normal.res.use = 0.5
ntiss10x.immune.r3 <- ntiss10x.immune.r3 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)

TSNEPlot(object = ntiss10x.immune.r3, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.r3, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.r3, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.r3, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.r3, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.immune.r3, do.return = TRUE, group.by = "compartment")
```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
DimPlot(ntiss10x.immune.r3, 1, 10, reduction.use = 'pca', group.by = 'channel')


```

```{r, fig.height=16, fig.width=10}

genes_to_check = toupper(c('CD68', 'MPEG1', 'PLD4', 'LCN2', 'HP', 'UBB', 'ACTB'))


FeaturePlot(ntiss10x.immune.r3, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r3.markers <- FindAllMarkers(ntiss10x.immune.r3, min.pct = 0.25, thresh.use = 0.25, only.pos = T, max.cells.per.ident = 200)

```


```{r, fig.height=200, fig.width=10}
genes_to_check <- ntiss10x.immune.r3.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.r3.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.r3, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters

```{r, fig.width=10, fig.height=10}
# All cells look relatively uniform, calling clusters in Round 2.
# There is bleed through from high-abundance transcripts like SFTPC. Example: Macrophages from 10X runs with more Type II cells have higher levels

```

# Subclustering normal immune cells (Round 4, Dendritic Cells)

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r4 = SubsetData(ntiss10x.immune.r2, ident.use = c(13,15,16))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r4 <- ntiss10x.immune.r4 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.r4, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r4, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r4, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.r4, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 10
normal.res.use = 1.5
ntiss10x.immune.r4 <- ntiss10x.immune.r4 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)

TSNEPlot(object = ntiss10x.immune.r4, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.r4, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.r4, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.r4, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.r4, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.immune.r4, do.return = TRUE, group.by = "compartment")
```


```{r, fig.height=45, fig.width=10}

genes_to_check = toupper(c('MARCO', 'Cd68', 'MSR1', 'MRC1', 'FCGR3A', 'CD14', 'CLEC10A', 'GPR183', 'BASP1', 'F13A1', 'SEPP1', 'RNASE1', 'LTF', 'LCN2', 'HP', 'CD24', 'CD19', 'MS4A1', 'CD3E', 'CD74', 'B2M'))


FeaturePlot(ntiss10x.immune.r4, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r4.markers <- FindAllMarkers(ntiss10x.immune.r4, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=70, fig.width=10}
genes_to_check <- ntiss10x.immune.r4.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.r4.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.r4, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters

```{r, fig.width=10, fig.height=10}

# There are based almost exclusively on the bulk RNAseq data from the Weissman lab. Some of the novel dendritic cells could also be macrophage subsets based on MACRO+ MSR+ MRC1+, but this is not likeily given these populations are also CD14+ and CD16+
# LILRA4+ SCT+ LRRC26+ GZMB+ -> Plasmacytoid Dendritic
# FCER1A+ CD1C+ PAK1+ PKIB+ -> Myeloid Dendritic Type 2
# FCER1A+ CD1C+ LAMP3+ CLEC9A+ -> Myeloid Dendritic Type 1
# CD79A+ CD27+ MS4A1- CD19- -> Plasma Cell
# GPR183+ EREG+ NAMPT+ CD14+ CD16+ -> EREG+ Dendritic
# GPR183+ IGSF21+ CD14+ CD16+ -> IGSF21+ Dendritic
# GPR183+ TREM2+ CD14+ CD16+ -> TREM2+ Dendritic


# stash current cluster IDs
ntiss10x.immune.r4 <- StashIdent(object = ntiss10x.immune.r4, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune.r4@ident)))
free_annotation <- c("Myeloid Dendritic Type 2",
                     #0
                     "TREM2+ Dendritic",
                     "IGSF21+ Dendritic",
                     "EREG+ Dendritic",
                     NA,
                     "Plasmacytoid Dendritic",
                     #5
                     "Plasma",
                     "Myeloid Dendritic Type 1")


ntiss10x.immune.r4@meta.data[,'free_annotation'] <- NA
ntiss10x.immune.r4@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune.r4@ident, from = cluster.ids, to = free_annotation))
```

# Subclustering normal immune cells (Round 5, Monocytes)

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r5 = SubsetData(ntiss10x.immune.r2, ident.use = c(13,15,9,3,10,7,16))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r5 <- ntiss10x.immune.r5 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.r5, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r5, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r5, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.r5, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 12
normal.res.use = 1.5
ntiss10x.immune.r5 <- ntiss10x.immune.r5 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)

TSNEPlot(object = ntiss10x.immune.r5, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.r5, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.r5, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.r5, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.r5, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.immune.r5, do.return = TRUE, group.by = "compartment")
```


```{r, fig.height=50, fig.width=10}

genes_to_check = toupper(c('MARCO', 'Cd68', 'MSR1', 'MRC1', 'FCGR3A', 'CD14', 'CLEC10A', 'GPR183', 'BASP1', 'F13A1', 'SEPP1', 'RNASE1', 'LTF', 'LCN2', 'HP', 'CD24', 'CD19', 'MS4A1', 'CD3E', 'CD74', 'B2M', 'CX3CR1'))


FeaturePlot(ntiss10x.immune.r5, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r5.markers <- FindAllMarkers(ntiss10x.immune.r5, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=110, fig.width=10}
genes_to_check <- ntiss10x.immune.r5.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.r5.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.r5, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}

# Dendritics were thrown in (clusters 5,6,8,9,10,11) to show differences to macrophages and monocytes
# LGALS2+ CD14+ NRG1+ S100A8+ S100A9+ S100A12+ CD14+ -> Classical Monocyte
# CDKN1C+ CD79B+ LYPD2+ CHST2+ IFITM2+ HES4+ CD16+ -> Nonclassical Monocyte

# stash current cluster IDs
ntiss10x.immune.r5 <- StashIdent(object = ntiss10x.immune.r5, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune.r5@ident)))
free_annotation <- c(NA,
                     #0
                     "Classical Monocyte",
                     "Classical Monocyte",
                     "Classical Monocyte",
                     "Nonclassical Monocyte",
                     NA,
                     #5
                     NA,
                     "Nonclassical Monocyte",
                     NA,
                     NA,
                     NA,
                     #10
                     NA)

ntiss10x.immune.r5@meta.data[,'free_annotation'] <- NA
ntiss10x.immune.r5@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune.r5@ident, from = cluster.ids, to = free_annotation))
```




# Combine all the annotations into the main object

```{r, fig.width=10, fig.height=10}
ntiss10x@meta.data[,'free_annotation'] <- NA

ntiss10x@meta.data[WhichCells(ntiss10x, ident = 30),'free_annotation'] <- 'Platelet/Megakaryocyte'


ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.epi@meta.data)),'free_annotation'] <- ntiss10x.epi@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.epi@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.epi.r2@meta.data)),'free_annotation'] <- ntiss10x.epi.r2@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.epi.r2@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.endo@meta.data)),'free_annotation'] <- ntiss10x.endo@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.endo@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.stromal@meta.data)),'free_annotation'] <- ntiss10x.stromal@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.stromal@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune@meta.data)),'free_annotation'] <- ntiss10x.immune@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r2@meta.data)),'free_annotation'] <- ntiss10x.immune.r2@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r2@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r4@meta.data)),'free_annotation'] <- ntiss10x.immune.r4@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r4@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r5@meta.data)),'free_annotation'] <- ntiss10x.immune.r5@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r5@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.Tcells@meta.data)),'free_annotation'] <- ntiss10x.immune.Tcells@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.Tcells@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.Tcells.r2@meta.data)),'free_annotation'] <- ntiss10x.immune.Tcells.r2@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.Tcells.r2@meta.data)),'free_annotation']

```


```{r, fig.width=10, fig.height=10}
TSNEPlot(ntiss10x, do.label = T, group.by = 'free_annotation', no.legend = T)

```

```{r, fig.width=10, fig.height=10}
sort(table(ntiss10x@meta.data[,'free_annotation']), decreasing = T)
```

# Patient specific DE tests of versus resident immune cells
```{r}
ntiss10x <- SetIdent(ntiss10x, cells.use = rownames(ntiss10x@meta.data[which(!is.na(ntiss10x@meta.data[,'free_annotation'])),]), ident.use =  ntiss10x@meta.data[which(!is.na(ntiss10x@meta.data[,'free_annotation'])),'free_annotation'])

#CD4Naive.markers <- FindMarkers(ntiss10x, ident.1 = 'CD4+ Naive T', ident.2 = 'CD4+ Naive T')
#CD4Nonnaive.markers <- FindMarkers(ntiss10x, ident.1 = 'CD4+ Nonnaive T', ident.2 = 'CD4+ Nonnaive T')
CD8Naive.markers <- FindMarkers(ntiss10x, ident.1 = 'CD8+ Naive T', ident.2 = 'CD8+ Naive T')
CD8Nonnaive.markers <- FindMarkers(ntiss10x, ident.1 = 'CD8+ Nonnaive T', ident.2 = 'CD8+ Nonnaive T')

NK.markers <- FindMarkers(ntiss10x, ident.1 = 'Natural Killer', ident.2 = 'Natural Killer')

ClassMono.markers <- FindMarkers(ntiss10x, ident.1 = 'Classical Monocyte', ident.2 = 'Classical Monocyte')
NonClassMono.markers <- FindMarkers(ntiss10x, ident.1 = 'Nonclassical Monocyte', ident.2 = 'Nonclassical Monocyte')

BPlasma.markers <- FindMarkers(ntiss10x, ident.1 = 'B', ident.2 = 'Plasma')


ntiss10x <- SetIdent(ntiss10x, cells.use = rownames(ntiss10x@meta.data[intersect(which(ntiss10x@meta.data[,'free_annotation'] == 'Plasmacytoid Dendritic'), which(ntiss10x@meta.data[,'tissue'] == 'Blood')),]), ident.use = 'Plasmacytoid Dendritic')


ntiss10x <- SetIdent(ntiss10x, cells.use = rownames(ntiss10x@meta.data[intersect(which(ntiss10x@meta.data[,'free_annotation'] == 'Plasmacytoid Dendritic'), which(ntiss10x@meta.data[,'tissue'] == 'Lung')),]), ident.use = 'Plasmacytoid Dendritic')

pDC.markers <- FindMarkers(ntiss10x, ident.1 = 'Plasmacytoid Dendritic', ident.2 = 'Plasmacytoid Dendritic')

#write.csv(CD4Naive.markers, file = "CD4Naive.markers.csv")
#write.csv(CD4Nonnaive.markers, file = "CD4Nonnaive.markers.csv")
write.csv(CD8Naive.markers, file = "CD8Naive.markers.csv")
write.csv(CD8Nonnaive.markers, file = "CD8Nonnaive.markers.csv")
write.csv(NK.markers, file = "NK.markers.csv")

write.csv(ClassMono.markers, file = "ClassMono.markers.csv")
write.csv(NonClassMono.markers, file = "NonClassMono.markers.csv")

write.csv(BPlasma.markers, file = "BPlasma.markers.csv")

write.csv(pDC.markers, file = "pDC.markers.csv")

```

# Removing suspected doublets/cells with low UMIs

```{r, fig.width=10, fig.height=10}
ntiss10x.anno = SubsetData(ntiss10x, cells.use = rownames(ntiss10x@meta.data[which(!is.na(ntiss10x@meta.data[,'free_annotation'])),]))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.anno <- ntiss10x.anno %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 40) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.anno, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.anno, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.anno, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.anno, pc.use = 28:36, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)


PCElbowPlot(ntiss10x.anno, num.pc = 40)
```


```{r, fig.width=10, fig.height=10}
n.pcs = 25
ntiss10x.anno <- ntiss10x.anno %>%
    RunTSNE(dims.use = 1:n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.anno, do.label = T, pt.size = 1.2, label.size = 4, group.by = 'free_annotation', no.legend = TRUE)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.anno, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.anno, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.anno, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.anno, do.return = TRUE, group.by = "location")
TSNEPlot(object = ntiss10x.anno, do.return = TRUE, group.by = "compartment")

```


```{r, fig.width=10, fig.height=10}

genes_to_check = toupper(c('Epcam', 'Cldn5', 'Ptprc', 'Col1a2'))

FeaturePlot(ntiss10x.anno, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

##Reassign clusters based on cell type


```{r, fig.width=10, fig.height=10}

clusters = 1:length(table(ntiss10x.anno@meta.data[,'free_annotation']))
names(clusters) = names(sort(table(ntiss10x.anno@meta.data[,'free_annotation']), decreasing = T))
ntiss10x.anno@meta.data[,'cluster_ids'] <- clusters[ntiss10x.anno@meta.data[,'free_annotation']]
clusters <- clusters[ntiss10x.anno@meta.data[,'free_annotation']]
names(clusters) <- rownames(ntiss10x.anno@meta.data)
ntiss10x.anno@ident <- as.factor(clusters)


```


```{r, fig.width=10, fig.height=10}
TSNEPlot(object = ntiss10x.anno, do.label = T, pt.size = 1.2, label.size = 4, no.legend = TRUE)

```


# Save the Robject for later
When you save the annotated tissue, please give it a name.

```{r, fig.width=10, fig.height=10}
ntiss10x.P1.anno <- ntiss10x.anno
filename = paste0("seurat/droplet_normal_lung_blood_seurat_ntiss10x.P1.anno.20190610.RC3.Robj")
print(filename)
save(ntiss10x.P1.anno, file=filename)
```

```{r, fig.width=10, fig.height=10}
# To reload a saved object
# load(file=filename)
```
