---
title: "The Human Lung Cell Atlas (10X)"
output:
  html_document: default
  html_notebook: default
---
# All Package versions used


```{r, fig.width=10, fig.height=10}
installed.packages()[,'Version']

```

# Specify the tissue of interest, run the boilerplate code which sets up the functions and environment, load the tissue object.


```{r, fig.width=10, fig.height=10}
tissue_of_interest = "all"
library(here)
source(here("boilerplate.R"))
tiss10x <- load_tissue_droplet(tissue_of_interest)
```


# Subclustering normal cells

```{r, fig.width=10, fig.height=10}
ntiss10x <- SubsetData(tiss10x, cells.use = rownames(tiss10x@meta.data[which(tiss10x@meta.data[,'region'] %in% c('Normal')),]))

```

```{r, fig.width=10, fig.height=10}
ntiss10x <- ntiss10x %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30 ) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 30
normal.res.use = 2
ntiss10x <- ntiss10x %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=15, fig.width=10}

genes_to_check = toupper(c('Pecam1', 'Epcam', 'Col1a1', 'Ptprc', 'Pdgfrb'))

FeaturePlot(ntiss10x, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

# Subclustering normal stromal cells

```{r, fig.width=10, fig.height=10}
 cells.to.use <- WhichCells(ntiss10x, ident = c(9,20,23))
 cells.to.use <- intersect(cells.to.use, intersect(names(which(ntiss10x@data['PECAM1',] < 0.1)),names(which(ntiss10x@data['GZMB',] < 0.1))))
 ntiss10x.stromal = SubsetData(ntiss10x, cells.use = cells.to.use)
```

```{r, fig.width=10, fig.height=10}
 ntiss10x.stromal <- ntiss10x.stromal %>% ScaleData() %>%
   FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
   RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r fig.height=10, fig.width=10}
 PCHeatmap(object = ntiss10x.stromal, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
 PCHeatmap(object = ntiss10x.stromal, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
 PCElbowPlot(ntiss10x.stromal)
```


```{r, fig.width=10, fig.height=10}
 normal.n.pcs = 9
 normal.res.use = 1.5
 ntiss10x.stromal <- ntiss10x.stromal %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
     resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
     RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

 TSNEPlot(object = ntiss10x.stromal, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
 # Batch and sample effects
 TSNEPlot(object = ntiss10x.stromal, do.return = TRUE, group.by = "channel", no.legend = T)
 TSNEPlot(object = ntiss10x.stromal, do.return = TRUE, group.by = "tissue")
 TSNEPlot(object = ntiss10x.stromal, do.return = TRUE, group.by = "region")
 TSNEPlot(object = ntiss10x.stromal, do.return = TRUE, group.by = "area")
 TSNEPlot(object = ntiss10x.stromal, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=60, fig.width=10}

 genes_to_check = toupper(c('Pecam1', 'Epcam', 'Col1a1', 'Ptprc', 'PDGFRA', 'PDGFRB', 'CNN1', 'KCNA5', 'LGR6', 'DES', 'TRPC6', 'HIGD1B', 'SLC38A5', 'GPC3', 'PI16', 'Serpinf1', 'SCARA5', 'THY1', 'PLIN2', 'APOE', 'HAS1', 'MEDAG', 'WT1', 'FGF18', 'ASPN', 'WIF1', 'MSLN', 'UPK3B'))

 FeaturePlot(ntiss10x.stromal, genes_to_check, pt.size = 2, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
 ntiss10x.stromal.markers <- FindAllMarkers(ntiss10x.stromal, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=150, fig.width=10}
 genes_to_check <- ntiss10x.stromal.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
 ntiss10x.stromal.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
 FeaturePlot(ntiss10x.stromal, genes_to_check$gene, pt.size = 2, nCol = 2, cols.use = my.cols, dark.theme = TRUE)
```

 ## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}
 #At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
 #Relying on lessons from mouse to call populations (a few representative markers are shown). All stromal cells are CD45- #CD31- EpCAM-
 #COX4I2+ HIGD1B+ GJA4+ TAGLN- CNN- -> Pericytes
 #ACTA2+ MYH11+ ASPN+ TYRP1+ -> Myofibroblasts


 # stash current cluster IDs
 ntiss10x.stromal <- StashIdent(object = ntiss10x.stromal, save.name = "cluster.ids")

 # enumerate current cluster IDs and the labels for them
 cluster.ids <- 0:max(as.numeric(levels(ntiss10x.stromal@ident)))
 free_annotation <- c("Pericyte",#0
                      "Pericyte",
                      "Pericyte",
                      'Pericyte',
                      NA,
                      NA,#5
                      NA,
                      'Pericyte',
                      NA,
                      NA,
                      "Myofibroblast",#10
                      NA,
                      NA)

 ntiss10x.stromal@meta.data[,'free_annotation'] <- NA
 ntiss10x.stromal@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.stromal@ident, from = cluster.ids, to = free_annotation))
```



# Subclustering normal stromal cells (Round 2, Fibroblasts)

```{r, fig.width=10, fig.height=10}
 cells.to.use <- WhichCells(ntiss10x.stromal, ident = c(5,6,11,12))
 ntiss10x.stromal.r2 = SubsetData(ntiss10x.stromal, cells.use = cells.to.use)
```

```{r, fig.width=10, fig.height=10}
 ntiss10x.stromal.r2 <- ntiss10x.stromal.r2 %>% ScaleData() %>%
   FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
   RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r fig.height=10, fig.width=10}
 PCHeatmap(object = ntiss10x.stromal.r2, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
 PCHeatmap(object = ntiss10x.stromal.r2, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
 PCElbowPlot(ntiss10x.stromal.r2)
```


```{r, fig.width=10, fig.height=10}
 normal.n.pcs = 6
 normal.res.use = 1.5
 ntiss10x.stromal.r2 <- ntiss10x.stromal.r2 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
     resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
     RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

 TSNEPlot(object = ntiss10x.stromal.r2, do.label = T, pt.size = 2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
 # Batch and sample effects
 TSNEPlot(object = ntiss10x.stromal.r2, do.return = TRUE, group.by = "channel", no.legend = T)
 TSNEPlot(object = ntiss10x.stromal.r2, do.return = TRUE, group.by = "tissue")
 TSNEPlot(object = ntiss10x.stromal.r2, do.return = TRUE, group.by = "region")
 TSNEPlot(object = ntiss10x.stromal.r2, do.return = TRUE, group.by = "area")
 TSNEPlot(object = ntiss10x.stromal.r2, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=70, fig.width=10}

 genes_to_check = toupper(c('Pecam1', 'Epcam', 'Col1a1', 'Ptprc', 'PDGFRA', 'PDGFRB', 'CNN1', 'KCNA5', 'LGR6', 'DES', 'TRPC6', 'HIGD1B', 'SLC38A5', 'FGFR4', 'ITGA8', 'GPC3', 'PI16', 'Serpinf1', 'SCARA5', 'THY1', 'SFRP2', 'PLIN2', 'APOE', 'WT1', 'TWIST1', 'MEDAG', 'FGF18', 'ASPN', 'WIF1', 'MSLN', 'UPK3B'))

 FeaturePlot(ntiss10x.stromal.r2, genes_to_check, pt.size = 2, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
 ntiss10x.stromal.r2.markers <- FindAllMarkers(ntiss10x.stromal.r2, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=80, fig.width=10}
 genes_to_check <- ntiss10x.stromal.r2.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
 ntiss10x.stromal.r2.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
 FeaturePlot(ntiss10x.stromal.r2, genes_to_check$gene, pt.size = 2, nCol = 2, cols.use = my.cols, dark.theme = TRUE)
```

 ## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}
 #At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
 #Relying on lessons from mouse to call populations (a few representative markers are shown). All stromal cells are CD45- #CD31- EpCAM-
 #SERPINF1+ PI16+ FBLN1+ SCARA5+ SLC38A5- -> Adventitial Fibroblast
 #SLC38A5+ GDF10+ GPC3+ ITGA8+ -> Alveolar Fibroblasts


 # stash current cluster IDs
 ntiss10x.stromal.r2 <- StashIdent(object = ntiss10x.stromal.r2, save.name = "cluster.ids")

 # enumerate current cluster IDs and the labels for them
 cluster.ids <- 0:max(as.numeric(levels(ntiss10x.stromal.r2@ident)))
 free_annotation <- c("Alveolar Fibroblast",
                      "Alveolar Fibroblast",
                      "Alveolar Fibroblast",
                      "Alveolar Fibroblast",
                      "Alveolar Fibroblast",
                      "Alveolar Fibroblast",
                      "Adventitial Fibroblast",
                      "Adventitial Fibroblast",
                      "Alveolar Fibroblast")

 ntiss10x.stromal.r2@meta.data[,'free_annotation'] <- NA
 ntiss10x.stromal.r2@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.stromal.r2@ident, from = cluster.ids, to = free_annotation))
```


# Subclustering normal stromal cells (Round 3, Smooth Muscle Cells)

```{r, fig.width=10, fig.height=10}
 cells.to.use <- WhichCells(ntiss10x.stromal, ident = c(4,8,9))
 ntiss10x.stromal.r3 = SubsetData(ntiss10x.stromal, cells.use = cells.to.use)
```

```{r, fig.width=10, fig.height=10}
 ntiss10x.stromal.r3 <- ntiss10x.stromal.r3 %>% ScaleData() %>%
   FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
   RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r fig.height=10, fig.width=10}
 PCHeatmap(object = ntiss10x.stromal.r3, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
 PCHeatmap(object = ntiss10x.stromal.r3, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
 PCElbowPlot(ntiss10x.stromal.r3)
```


```{r, fig.width=10, fig.height=10}
 normal.n.pcs = 6
 normal.res.use = 1
 ntiss10x.stromal.r3 <- ntiss10x.stromal.r3 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
     resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
     RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

 TSNEPlot(object = ntiss10x.stromal.r3, do.label = T, pt.size = 2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
 # Batch and sample effects
 TSNEPlot(object = ntiss10x.stromal.r3, do.return = TRUE, group.by = "channel", no.legend = T)
 TSNEPlot(object = ntiss10x.stromal.r3, do.return = TRUE, group.by = "tissue")
 TSNEPlot(object = ntiss10x.stromal.r3, do.return = TRUE, group.by = "region")
 TSNEPlot(object = ntiss10x.stromal.r3, do.return = TRUE, group.by = "area")
 TSNEPlot(object = ntiss10x.stromal.r3, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=70, fig.width=10}

 genes_to_check = toupper(c('Pecam1', 'Epcam', 'Col1a1', 'Ptprc', 'PDGFRA', 'PDGFRB', 'CNN1', 'KCNA5', 'LGR6', 'DES', 'TRPC6', 'HIGD1B', 'SLC38A5', 'FGFR4', 'ITGA8', 'GPC3', 'PI16', 'Serpinf1', 'SCARA5', 'THY1', 'SFRP2', 'PLIN2', 'APOE', 'WT1', 'TWIST1', 'MEDAG', 'FGF18', 'ASPN', 'WIF1', 'MSLN', 'UPK3B'))

 FeaturePlot(ntiss10x.stromal.r3, genes_to_check, pt.size = 2, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
 ntiss10x.stromal.r3.markers <- FindAllMarkers(ntiss10x.stromal.r3, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=80, fig.width=10}
 genes_to_check <- ntiss10x.stromal.r3.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
 ntiss10x.stromal.r3.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
 FeaturePlot(ntiss10x.stromal.r3, genes_to_check$gene, pt.size = 2, nCol = 2, cols.use = my.cols, dark.theme = TRUE)
```

## Assigning cell type identity to clusters

```{r, fig.width=10, fig.height=10}
 #At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
 #Relying on lessons from mouse to call populations (a few representative markers are shown). All stromal cells are CD45- #CD31- EpCAM-

 #ACTA2+ TAGLN+ CNN1+ MYH11+ DES+ LGR6+ KCN5A+ -> Airway Smooth Muscle Cell
 #ACTA2+ TAGLN+ CNN1+ MYH11+ ADAMTS4 -> Vascular Smooth Muscle Cell



 # stash current cluster IDs
 ntiss10x.stromal.r3 <- StashIdent(object = ntiss10x.stromal.r3, save.name = "cluster.ids")

 # enumerate current cluster IDs and the labels for them
 cluster.ids <- 0:max(as.numeric(levels(ntiss10x.stromal.r3@ident)))
 free_annotation <- c('Airway Smooth Muscle',
                      'Vascular Smooth Muscle',
                      'Airway Smooth Muscle',
                      'Airway Smooth Muscle')

 ntiss10x.stromal.r3@meta.data[,'free_annotation'] <- NA
 ntiss10x.stromal.r3@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.stromal.r3@ident, from = cluster.ids, to = free_annotation))
```


# Subclustering normal epithelial cells

```{r, fig.width=10, fig.height=10}
ntiss10x.epi = SubsetData(ntiss10x, ident.use = c(21,29,22,26,13,29,30))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi <- ntiss10x.epi %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.epi, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.epi, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCElbowPlot(ntiss10x.epi)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 9
normal.res.use = 1
ntiss10x.epi <- ntiss10x.epi %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.epi, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.epi, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.epi, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.epi, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.epi, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.epi, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=50, fig.width=10}

genes_to_check = toupper(c('Pecam1', 'Epcam', 'Col1a1', 'Ptprc', 'Muc5ac', 'Muc5b', 'SFTPC', 'ETV5', 'hhip','WIF1', 'AXIN2', 'FOXJ1', 'ltf', 'SCGB1A1', 'SCGB3A2', 'WIF1', 'HHIP', 'SERPINA1', 'KRT5', 'NGFR', 'KRT14', 'Muc12'))

FeaturePlot(ntiss10x.epi, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = T)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.markers <- FindAllMarkers(ntiss10x.epi, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=111, fig.width=10}
genes_to_check <- ntiss10x.epi.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.epi.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.epi, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

# Subclustering normal epithelial cells (Round 2)

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r2 = SubsetData(ntiss10x.epi, ident.use = c(0,2,4,5,6,7,10))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r2 <- ntiss10x.epi.r2 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.epi.r2, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.epi.r2, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCElbowPlot(ntiss10x.epi.r2)
```


```{r, fig.width=5, fig.height=5}
normal.n.pcs = 8
normal.res.use = 1
ntiss10x.epi.r2 <- ntiss10x.epi.r2 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.epi.r2, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.epi.r2, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.epi.r2, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.epi.r2, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.epi.r2, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.epi.r2, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=70, fig.width=10}

genes_to_check = toupper(c('Pecam1', 'Epcam', 'Col1a1', 'Acta2', 'Ptprc', 'Muc5ac', 'Muc5b', 'hes1', 'mki67', 'krt7', 'myc', 'FOXJ1', 'TUBB4B', 'SCGB1A1','KRT5', 'NGFR', 'SPRR3', 'Hes1', 'KRT14', 'Ascl3', 'Chga', 'Muc12'))

FeaturePlot(ntiss10x.epi.r2, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = T)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r2.markers <- FindAllMarkers(ntiss10x.epi.r2, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=20, fig.width=10}
genes_to_check <- ntiss10x.epi.r2.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.epi.r2.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.epi.r2, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
VlnPlot(ntiss10x.epi.r2, c('SPDEF','MUC5AC','MUC5B','SPRR1A','CFTR','SCGB1A1','SCGB3A2'))
```


# Subclustering normal epithelial cells (Round 3)

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r3 = SubsetData(ntiss10x.epi.r2, ident.use = c(0,1,2,3,4,5,7))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r3 <- ntiss10x.epi.r3 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.epi.r3, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.epi.r3, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCElbowPlot(ntiss10x.epi.r3)
```


```{r, fig.width=5, fig.height=5}
normal.n.pcs = 8
normal.res.use = 1
ntiss10x.epi.r3 <- ntiss10x.epi.r3 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.epi.r3, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.epi.r3, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.epi.r3, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.epi.r3, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.epi.r3, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.epi.r3, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=70, fig.width=10}

genes_to_check = toupper(c('Pecam1', 'Epcam',  'Ptprc', 'Muc5ac', 'Muc5b', 'muc12', 'muc7','muc16','muc22', 'FOXJ1', 'TUBB4B', 'SCGB1A1','scgb3a2','pigr','KRT5', 'NGFR', 'krt7', 'Krt13', 'Postn', 'Myc', 'Hes1', 'CYP2F1'))

FeaturePlot(ntiss10x.epi.r3, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = T)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r3.markers <- FindAllMarkers(ntiss10x.epi.r3, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=100, fig.width=10}
genes_to_check <- ntiss10x.epi.r3.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.epi.r3.markers %>% group_by(cluster) %>% top_n(-20, p_val_adj)
FeaturePlot(ntiss10x.epi.r3, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}
# At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
# There is still considerable heterogeneity among the cells, but we don't have enough cells to go further with Seurat
# KRT5+ KRT14+ -> Basal cells
# SCGB1A1+ SCGB3A2+ SP*- FOXJ1- -> Club Cells
# FOXJ1+ TUBB4+ -> Ciliated Cell
# MUC5B+ SPEDF+ MUC5AC- -> Mucous Cell

# stash current cluster IDs
ntiss10x.epi.r3 <- StashIdent(object = ntiss10x.epi.r3, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.epi.r3@ident)))
free_annotation <- c("Ciliated",
                     'Ciliated',
                     'Mucous',
                     'Basal',
                     'Ciliated',
                     NA,
                     'Mucous',
                     'Club',
                     'Ciliated')

ntiss10x.epi.r3@meta.data[,'free_annotation'] <- NA
ntiss10x.epi.r3@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.epi.r3@ident, from = cluster.ids, to = free_annotation))
TSNEPlot(ntiss10x.epi.r3, group.by = "free_annotation", do.label = TRUE)
```

#Epithelial Round 4 (Alveolar cells)

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r4 = SubsetData(ntiss10x.epi, ident.use = c(9,3,8,1))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r4 <- ntiss10x.epi.r4 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.epi.r4, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.epi.r4, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCElbowPlot(ntiss10x.epi.r4)
```


```{r, fig.width=5, fig.height=5}
normal.n.pcs = 7
normal.res.use = 1
ntiss10x.epi.r4 <- ntiss10x.epi.r4 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.epi.r4, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.epi.r4, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.epi.r4, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.epi.r4, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.epi.r4, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.epi.r4, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=10, fig.width=10}

genes_to_check = toupper(c('sftpc', 'hhip',  'wif1','ca2', 'tcf7l2', 'ager'))

FeaturePlot(ntiss10x.epi.r4, genes_to_check, pt.size = 1, nCol = 3, cols.use = my.cols, dark.theme = T)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.epi.r4.markers <- FindAllMarkers(ntiss10x.epi.r4, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=50, fig.width=10}
genes_to_check <- ntiss10x.epi.r4.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.epi.r4.markers %>% group_by(cluster) %>% top_n(-20, p_val_adj)
FeaturePlot(ntiss10x.epi.r4, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}
# At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
# SFP*+ ETV5+ MUC1+ -> Alveolar Epithelial Type 2 Cell
# HOPX+ PDPN+ CLIC5+ SFP*- -> Alveolar Epithelial Type 1 Cell
# Above, plus HHIP- WIF1- CA2-

# stash current cluster IDs
ntiss10x.epi.r4 <- StashIdent(object = ntiss10x.epi.r4, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.epi.r4@ident)))
free_annotation <- c("Alveolar Epithelial Type 2",
                     'Alveolar Epithelial Type 1',
                     'Alveolar Epithelial Type 1',
                     'Alveolar Epithelial Type 2',
                     'Alveolar Epithelial Type 2',
                     'Alveolar Epithelial Type 2',
                     'Alveolar Epithelial Type 1',
                     'Alveolar Epithelial Type 2')

ntiss10x.epi.r4@meta.data[,'free_annotation'] <- NA
ntiss10x.epi.r4@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.epi.r4@ident, from = cluster.ids, to = free_annotation))



```




# Subclustering normal endothelial cells

```{r, fig.width=10, fig.height=10}
cells.to.use = WhichCells(ntiss10x, ident = c(0,2,3,8,10,12,15,16,25,27,28))
cells.to.use = intersect(intersect(cells.to.use, names(which(ntiss10x@data['PTPRC',cells.to.use] < 0.1))), names(which(ntiss10x@data['EPCAM',cells.to.use] < 0.1)))


ntiss10x.endo = SubsetData(ntiss10x, cells.use = cells.to.use)
```


```{r, fig.width=10, fig.height=10}
ntiss10x.endo <- ntiss10x.endo %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.endo, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.endo, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCElbowPlot(ntiss10x.endo)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 8
normal.res.use = 1
ntiss10x.endo <- ntiss10x.endo %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.endo, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.endo, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.endo, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.endo, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.endo, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.endo, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=45, fig.width=10}

genes_to_check = toupper(c('Pecam1', 'Epcam', 'Col1a1', 'Ptprc', 'Cxcl12', 'Dkk2', 'Gja5', 'Slc6a4', 'Gpihbp1', 'Aplnr', 'Apln', 'Ednrb', 'Myc', 'Spry1', 'Ackr1', 'Rspo3', 'Ptgis', 'Prox1', 'PDPN', 'CTNNB1', 'FOS'))

FeaturePlot(ntiss10x.endo, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.endo.markers <- FindAllMarkers(ntiss10x.endo, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=80, fig.width=10}
genes_to_check <- ntiss10x.endo.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.endo.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.endo, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}
#
# At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
# Relying on lessons from mouse to call populations (a few representative markers are shown). All endothelial cells are CD45- EpCAM- Col1A1-
# IL7R+ SLC6A4+ FCN3+ C7- CPE- ACKR1- -> Capillary Type 1 Cell
# APLN+ EDNRB+ HPGD+ IL1RL1+  -> Capillary Type 2 Cell
# CXCL12+ DKK2+ GJA5+ SERPINE2+ -> Artery Cell
# RSPO+ ACKR1+ PTGIS+ PTDGS+ CPE+ -> Vein Cell
# TFF3+ PDPN+ CCL21+ IGF1+ -> Lymphatic Cell
# Capillary Intermediates have mixed expression of Type 1 and 2 markers


# stash current cluster IDs
ntiss10x.endo <- StashIdent(object = ntiss10x.endo, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.endo@ident)))
free_annotation <- c("Capillary 1",
                     "Capillary 1",
                     "Capillary 2",
                     "Capillary 1",
                     "Capillary 1",
                     "Capillary 2",
                     "Artery",
                     "Capillary 2",
                     "Vein",
                     "Capillary Intermediate 1",
                     "Capillary Intermediate 2",
                     "Lymphatics")

ntiss10x.endo@meta.data[,'free_annotation'] <- NA
ntiss10x.endo@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.endo@ident, from = cluster.ids, to = free_annotation))
TSNEPlot(ntiss10x.endo, group.by = "free_annotation", do.label = TRUE)
```

# Subclustering normal immune cells

```{r, fig.width=10, fig.height=10}
ntiss10x.immune = SubsetData(ntiss10x, ident.use = c(4,33,34,36,1,17,5,6,7,32,31,37,24,19,17,14,35,11,18))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune <- ntiss10x.immune %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 20
normal.res.use = 1
ntiss10x.immune <- ntiss10x.immune %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)

TSNEPlot(object = ntiss10x.immune, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.immune, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=50, fig.width=10}

genes_to_check = toupper(c('PTPRC', 'PECAM1', 'EPCAM', 'COL1A1', 'CD3E', 'KLRB1', 'CD79A', 'CD19', 'MS4A1', 'CD24', 'CD68', 'FCGR3A', 'CD14', 'MARCO', 'MSR1', 'MRC1', 'GPR183', 'CPA3', 'MKI67'))


FeaturePlot(ntiss10x.immune, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.markers <- FindAllMarkers(ntiss10x.immune, min.pct = 0.25, thresh.use = 0.25, only.pos = T, max.cells.per.ident = 1000)

```


```{r, fig.height=170, fig.width=10}
genes_to_check <- ntiss10x.immune.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}
# At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
# Calling immune cells based on canonical markers and bulk RNAseq data from the Weissman Lab (Rahul said this could be made available, if needed). All immune cells are EpCAM- Col1a1-, some are CD31+

# MARCO+ MRC1+ MSR1+ CD68 CD74+ CD14- FCGR3A- -> Macrophage
# CPA3+ MS4A2+ RGS13+ GATA2+ KIT+ -> Basophil/Mast Cell Type 1/2

# stash current cluster IDs
ntiss10x.immune <- StashIdent(object = ntiss10x.immune, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune@ident)))
free_annotation <- c(NA,
                     'Macrophage',
                     'Macrophage',
                     'Macrophage',
                     'Macrophage',
                     NA,#5
                     'Mast Cell/Basophil Type 1',
                     NA,
                     NA,
                     NA,
                     NA,#10
                     NA,
                     NA,
                     NA,
                     NA,
                     NA,#15
                     'Mast Cell/Basophil Type 2',
                     NA,
                     NA)

ntiss10x.immune@meta.data[,'free_annotation'] <- NA
ntiss10x.immune@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune@ident, from = cluster.ids, to = free_annotation))
```

# Subclustering normal immune cells (T cells)

```{r, fig.width=10, fig.height=10}
cells.to.use <- WhichCells(ntiss10x.immune, ident = c(0,7,8,9,17,13))
cells.to.use = intersect(cells.to.use, names(which(ntiss10x.immune@data['RAMP2',cells.to.use] < 0.1)))
ntiss10x.immune.Tcells = SubsetData(ntiss10x.immune, cells.use = cells.to.use)
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells <- ntiss10x.immune.Tcells %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.Tcells, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.Tcells, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.Tcells, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.Tcells, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 9
normal.res.use = 0.5
ntiss10x.immune.Tcells <- ntiss10x.immune.Tcells %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)

TSNEPlot(object = ntiss10x.immune.Tcells, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.Tcells, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.Tcells, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.Tcells, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.Tcells, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.immune.Tcells, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=60, fig.width=10}

genes_to_check = toupper(c('CD79A', 'CD24', 'CD19', 'MS4A1', 'PTPRC', 'PECAM1', 'EPCAM', 'COL1A1', 'CD3E', 'CD4', 'CD8A', 'KLRB1','CHST2', 'MKI67', 'FCER1G', 'TYROBP', 'AREG', 'HOPX', 'LTB', 'COTL1', 'LDHB', 'CCR7', 'LEF1', 'GZMH', 'GZMB', 'IL7R', 'GZMK', 'DUSP2', 'CXCR4', 'DUSP1'))

FeaturePlot(ntiss10x.immune.Tcells, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells.markers <- FindAllMarkers(ntiss10x.immune.Tcells, min.pct = 0.25, thresh.use = 0.25, only.pos = T, max.cells.per.ident = 500)

```


```{r, fig.height=60, fig.width=10}
genes_to_check <- ntiss10x.immune.Tcells.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.Tcells.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.Tcells, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}
# At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
# FCER1G+ TYROBP+ GZMB+ CHST2+ HOPX+ CD3E- -> Natural Killer Cell
# MKI67+ TOP2A+ STM1+ CD3E(mixed) FCER1G(mixed) -> Proliferating NK/T Cell
# CD79A+ CD24+ MS4A1+ CD19+ -> B Cell

# stash current cluster IDs
ntiss10x.immune.Tcells <- StashIdent(object = ntiss10x.immune.Tcells, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune.Tcells@ident)))
free_annotation <- c('Itinerant Natural Killer',
                     'Itinerant Natural Killer',
                     NA,
                     NA,
                     NA,
                     #5
                     'Proliferating NK/T',
                     "B")

ntiss10x.immune.Tcells@meta.data[,'free_annotation'] <- NA
ntiss10x.immune.Tcells@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune.Tcells@ident, from = cluster.ids, to = free_annotation))
```



# Subclustering normal immune cells (T Cells, Round 2)

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells.r2 = SubsetData(ntiss10x.immune.Tcells, ident.use = c(2,3,4))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells.r2 <- ntiss10x.immune.Tcells.r2 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.Tcells.r2, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.Tcells.r2, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.Tcells.r2, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.Tcells.r2, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 9
normal.res.use = 1.5
ntiss10x.immune.Tcells.r2 <- ntiss10x.immune.Tcells.r2 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
      RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)


TSNEPlot(ntiss10x.immune.Tcells.r2, label.size = 4, pt.size = 1.2, do.label = T)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.Tcells.r2, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.Tcells.r2, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.Tcells.r2, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.Tcells.r2, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.immune.Tcells.r2, do.return = TRUE, group.by = "compartment")
```

```{r, fig.height=50, fig.width=10}

genes_to_check = toupper(c('PTPRC', 'PECAM1', 'EPCAM', 'COL1A1', 'CD3E', 'CD4', 'CD8A', 'KLRB1','CHST2', 'MKI67', 'FCER1G', 'TYROBP', 'AREG', 'HOPX', 'LTB', 'COTL1', 'LDHB', 'CCR7', 'LEF1', 'GZMH', 'GZMB', 'IL7R', 'GZMK', 'DUSP2', 'CXCR4', 'DUSP1', 'FKBP5'))

FeaturePlot(ntiss10x.immune.Tcells.r2, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.Tcells.r2.markers <- FindAllMarkers(ntiss10x.immune.Tcells.r2, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=80, fig.width=10}
genes_to_check <- ntiss10x.immune.Tcells.r2.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.Tcells.r2.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.Tcells.r2, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)
```

## Assigning cell type identity to clusters


```{r, fig.width=10, fig.height=10}
# At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
# CD3E+ CD8+ GZMH+ GZMB+ LTB- IL7R- GZMK- DUSP2- -> CD8+ Naive T Cell
# CD3E+ CD8+ GZMK+ KLRB1+ IL7R+ DUSP2+ GZMH- GZMB- -> CD8+ Memory T Cell
# CD3E+ CD4+ LTB+ COTL1+ LDHB+ CCR7- LEF1- -> CD4+ Memory T Cell
# CD3E+ CD4+ LTB+ LDHB+ CCR7+ LEF1+ COTL1- -> CD4+ Naive T Cell
# CD3E+ CD8+ FCER1G+ TYROBP+ -> Natural Killer T Cell



# stash current cluster IDs
ntiss10x.immune.Tcells.r2 <- StashIdent(object = ntiss10x.immune.Tcells.r2, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune.Tcells.r2@ident)))
free_annotation <- c("Itinerant CD8+ Naive T",
                     "Itinerant CD4+ Memory T",
                     "Itinerant CD8+ Memory T",
                     "Itinerant CD8+ Memory T",
                     "Itinerant CD4+ Memory T",
                     "Itinerant CD8+ Naive T",
                     "Itinerant CD4+ Memory T",
                     "Itinerant Natural Killer T",
                     "Itinerant CD8+ Naive T",
                     "Itinerant CD4+ Naive T")

ntiss10x.immune.Tcells.r2@meta.data[,'free_annotation'] <- NA
ntiss10x.immune.Tcells.r2@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune.Tcells.r2@ident, from = cluster.ids, to = free_annotation))
```


# Subclustering normal immune cells (Round 2, Non-T Cells)

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r2 = SubsetData(ntiss10x.immune, ident.use = c(5,10,11,12,14,15,18))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r2 <- ntiss10x.immune.r2 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.r2, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r2, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r2, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.r2, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 11
normal.res.use = 1
ntiss10x.immune.r2 <- ntiss10x.immune.r2 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=20)

TSNEPlot(object = ntiss10x.immune.r2, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.r2, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.r2, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.r2, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.r2, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.immune.r2, do.return = TRUE, group.by = "compartment")
```


```{r, fig.height=70, fig.width=10}

genes_to_check = toupper(c('PECAM1', 'TMEM100', 'CLDN5', 'RAMP2', 'CD3E', 'CD8A', 'CD4', 'CD79A', 'CD27', 'GPR183', 'MRC1', 'MARCO', 'KLRB1', 'S100A8',
                           'FCER1A','CD1C','PAK1',#MYELOID DENDRITIC CELL
                           'LGALS2','CD14','NRG1', #CLASSOCIAL MONOCYTE
                           'FCGR3A', 'LAMP3','TIMP3', 'IGSF21', 'NAMPT', 'TREM2', 'MKI67', 'TOP2A'))


FeaturePlot(ntiss10x.immune.r2, genes_to_check, pt.size = 2, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r2.markers <- FindAllMarkers(ntiss10x.immune.r2, min.pct = 0.25, thresh.use = 0.25, only.pos = T, max.cells.per.ident = 200)

```


```{r, fig.height=100, fig.width=10}
genes_to_check <- ntiss10x.immune.r2.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.r2.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.r2, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)
```

## Assigning cell type identity to clusters

```{r, fig.width=10, fig.height=10}
# At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
# MARCO+ MRC1+ MSR1+ CD68 CD74+ CD14- FCGR3A- -> Macrophage
# MKI67+ TOP2A+ STM1+ MARCO+ MRC1+ MSR1+ CD68 CD74+ CD14- FCGR3A- -> Proliferating Macrophage
# LILRA4+ SCT+ LRRC26+ GZMB+ -> Plasmacytoid Dendritic Cell

# stash current cluster IDs
ntiss10x.immune.r2 <- StashIdent(object = ntiss10x.immune.r2, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune.r2@ident)))
free_annotation <- c(NA,#0
                     NA,
                     NA,
                     "Macrophage",
                     NA,
                     NA,#5
                     NA,
                     NA,
                     NA,
                     'Proliferating Macrophage',
                     NA,#10
                     NA,
                     "Plasmacytoid Dendritic")

ntiss10x.immune.r2@meta.data[,'free_annotation'] <- NA
ntiss10x.immune.r2@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune.r2@ident, from = cluster.ids, to = free_annotation))
```


# Subclustering normal immune cells (Round 3, Dendritic and Monocytes)

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r3 = SubsetData(ntiss10x.immune.r2, ident.use = c(2,7,8))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r3 <- ntiss10x.immune.r3 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.r3, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r3, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r3, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.r3, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 6
normal.res.use = 2
ntiss10x.immune.r3 <- ntiss10x.immune.r3 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE, prune.SNN = 1/3, plot.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.immune.r3, do.label = T, pt.size = 3, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.r3, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.r3, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.r3, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.r3, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.immune.r3, do.return = TRUE, group.by = "compartment")
```


```{r, fig.height=60, fig.width=10}

genes_to_check = toupper(c('CD3E', 'CD8A', 'CD4', 'CD79A', 'CD27', 'GPR183', 'MRC1', 'MARCO', 'KLRB1',
                           'FCER1A','CD1C','PAK1',#MYELOID DENDRITIC CELL
                           'LGALS2','CD14','NRG1', #CLASSOCIAL MONOCYTE
                           'FCGR3A', 'LAMP3','TIMP3', 'IGSF21', 'NAMPT', 'TREM2', 'MKI67', 'TOP2A',
                           'S100A8', 'S100A9', 'IL1B', 'CXCR4', 'DUSP1', 'FKBP5'))

FeaturePlot(ntiss10x.immune.r3, genes_to_check, pt.size = 2, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r3.markers <- FindAllMarkers(ntiss10x.immune.r3, min.pct = 0.25, thresh.use = 0.25, only.pos = T, max.cells.per.ident = 200)

```


```{r, fig.height=90, fig.width=10}
genes_to_check <- ntiss10x.immune.r3.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.r3.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.r3, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)
```

## Assigning cell type identity to clusters

```{r, fig.width=10, fig.height=10}
# At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
# There are based almost exclusively on the bulk RNAseq data from the Weissman lab. Some of these cells could also be macrophage subsets based on MACRO+ MSR+ MRC1+, but this is not likeily given these populations are also CD14+ and CD16+
# FCER1A+ CD1C+ PAK1+ PKIB+ -> Myeloid Dendritic Type 2 Cell
# FCER1A+ CD1C+ LAMP3+ CLEC9A+ -> Myeloid Dendritic Type 1 Cell
# LGALS2+ CD14+ NRG1+ S100A8+ S100A9+ S100A12+ CD14+ -> Classical Monocytes
# GPR183+ EREG+ NAMPT+ CD14+ CD16+ -> EREG+ Dendritic Cell
# GPR183+ IGSF21+ CD14+ CD16+ -> IGSF21+ Dendritic Cell



# stash current cluster IDs
ntiss10x.immune.r3 <- StashIdent(object = ntiss10x.immune.r3, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune.r3@ident)))
free_annotation <- c("EREG+ Dendritic",#0
                     "IGSF21+ Dendritic",
                     "IGSF21+ Dendritic",
                     "EREG+ Dendritic",
                     "Itinerant Classical Monocyte",
                     "Myeloid Dendritic Type 1",
                     "Itinerant Classical Monocyte",
                     "Itinerant Classical Monocyte",
                     "Itinerant Classical Monocyte",
                     "Myeloid Dendritic Type 2",
                     "Myeloid Dendritic Type 1")


ntiss10x.immune.r3@meta.data[,'free_annotation'] <- NA
ntiss10x.immune.r3@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune.r3@ident, from = cluster.ids, to = free_annotation))
```

# Subclustering normal immune cells (Round 4, Monocytes)

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r4 = SubsetData(ntiss10x.immune.r2, ident.use = c(0,1,4,5))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r4 <- ntiss10x.immune.r4 %>% ScaleData() %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 30) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.immune.r4, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r4, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.immune.r4, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)

PCElbowPlot(ntiss10x.immune.r4, num.pc = 30)
```


```{r, fig.width=10, fig.height=10}
normal.n.pcs = 7
normal.res.use = 1
ntiss10x.immune.r4 <- ntiss10x.immune.r4 %>% FindClusters(reduction.type = "pca", dims.use = 1:normal.n.pcs,
    resolution = normal.res.use, print.output = 0, save.SNN = TRUE, force.recalc = TRUE) %>%
    RunTSNE(dims.use = 1:normal.n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.immune.r4, do.label = T, pt.size = 1.2, label.size = 4)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.immune.r4, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.immune.r4, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.immune.r4, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.immune.r4, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.immune.r4, do.return = TRUE, group.by = "compartment")
```


```{r, fig.height=70, fig.width=10}

genes_to_check = toupper(c('PTPRC', 'COL1A1', 'PECAM1', 'EPCAM','CD3E', 'CD8A', 'CD4', 'CD79A', 'CD27', 'GPR183', 'MRC1', 'MARCO', 
                           'KLRB1', 'FCER1A','CD1C','PAK1',#MYELOID DENDRITIC CELL
                           'LGALS2','CD14','NRG1', 'OLR1', #CLASSOCIAL MONOCYTE
                           'FCGR3A', 'LAMP3','TIMP3', 'IGSF21', 'NAMPT', 'TREM2', 'MKI67', 'TOP2A',
                           'S100A8', 'S100A9', 'IL1B', 'CXCR4', 'DUSP1', 'FKBP5'))


FeaturePlot(ntiss10x.immune.r4, genes_to_check, pt.size = 2, nCol = 2, cols.use = my.cols, dark.theme = TRUE)

```

```{r, fig.width=10, fig.height=10}
ntiss10x.immune.r4.markers <- FindAllMarkers(ntiss10x.immune.r4, min.pct = 0.25, thresh.use = 0.25, only.pos = T)

```


```{r, fig.height=40, fig.width=10}
genes_to_check <- ntiss10x.immune.r4.markers %>% group_by(cluster) %>% top_n(-4, p_val_adj)
ntiss10x.immune.r4.markers %>% group_by(cluster) %>% top_n(-10, p_val_adj)
FeaturePlot(ntiss10x.immune.r4, genes_to_check$gene, pt.size = 1, nCol = 2, cols.use = my.cols, dark.theme = TRUE)
```

## Assigning cell type identity to clusters

```{r, fig.width=10, fig.height=10}
# At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:
# There are based almost exclusively on the bulk RNAseq data from the Weissman lab. Some of these cells could also be macrophage subsets based on MACRO+ MSR+ MRC1+, but this is not likeily given these populations are also CD14+ and CD16+
# LGALS2+ CD14+ NRG1+ S100A8+ S100A9+ S100A12+ CD14+ -> Classical Monocytes
# CDKN1C+ CD79B+ LYPD2+ CHST2+ IFITM2+ HES4+ CD16+ -> Nonclassical Monocytes
# Intermediate monocytes have marker gene expression of both classical and nonclassical monocytes.


# stash current cluster IDs
ntiss10x.immune.r4 <- StashIdent(object = ntiss10x.immune.r4, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- 0:max(as.numeric(levels(ntiss10x.immune.r4@ident)))
free_annotation <- c("Intermediate Monocyte",#0
                     "Itinerant Classical Monocyte",
                     "Itinerant Nonclassical Monocyte",
                     "OLR1+ Classical Monocyte",
                     "Itinerant Nonclassical Monocyte",
                     "OLR1+ Classical Monocyte")


ntiss10x.immune.r4@meta.data[,'free_annotation'] <- NA
ntiss10x.immune.r4@meta.data[,'free_annotation'] <- as.character(plyr::mapvalues(x = ntiss10x.immune.r4@ident, from = cluster.ids, to = free_annotation))
```


# Combine all the annotations into the main object

```{r, fig.width=10, fig.height=10}
ntiss10x@meta.data[,'free_annotation'] <- NA

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.epi.r3@meta.data)),'free_annotation'] <- ntiss10x.epi.r3@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.epi.r3@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.epi.r4@meta.data)),'free_annotation'] <- ntiss10x.epi.r4@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.epi.r4@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.endo@meta.data)),'free_annotation'] <- ntiss10x.endo@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.endo@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.stromal@meta.data)),'free_annotation'] <- ntiss10x.stromal@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.stromal@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.stromal.r2@meta.data)),'free_annotation'] <- ntiss10x.stromal.r2@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.stromal.r2@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.stromal.r3@meta.data)),'free_annotation'] <- ntiss10x.stromal.r3@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.stromal.r3@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune@meta.data)),'free_annotation'] <- ntiss10x.immune@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.Tcells@meta.data)),'free_annotation'] <- ntiss10x.immune.Tcells@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.Tcells@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.Tcells.r2@meta.data)),'free_annotation'] <- ntiss10x.immune.Tcells.r2@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.Tcells.r2@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r2@meta.data)),'free_annotation'] <- ntiss10x.immune.r2@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r2@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r3@meta.data)),'free_annotation'] <- ntiss10x.immune.r3@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r3@meta.data)),'free_annotation']

ntiss10x@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r4@meta.data)),'free_annotation'] <- ntiss10x.immune.r4@meta.data[intersect(rownames(ntiss10x@meta.data[which(is.na(ntiss10x@meta.data[,'free_annotation'])),]),rownames(ntiss10x.immune.r4@meta.data)),'free_annotation']

```


```{r, fig.width=10, fig.height=10}
TSNEPlot(ntiss10x, do.label = T, group.by = 'free_annotation', no.legend = T)

```

```{r, fig.width=10, fig.height=10}
sort(table(ntiss10x@meta.data[,'free_annotation']), decreasing = T)
```

# Removing suspected doublets/cells with low UMIs

```{r, fig.width=10, fig.height=10}
ntiss10x.anno = SubsetData(ntiss10x, cells.use = rownames(ntiss10x@meta.data[which(!is.na(ntiss10x@meta.data[,'free_annotation'])),]))
```

```{r, fig.width=10, fig.height=10}
ntiss10x.anno <- ntiss10x.anno %>%
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE, pcs.compute = 40) %>% ProjectPCA(do.print = FALSE)
```

```{r, fig.width=10, fig.height=10}
PCHeatmap(object = ntiss10x.anno, pc.use = 1:9, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.anno, pc.use = 10:18, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.anno, pc.use = 19:27, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)
PCHeatmap(object = ntiss10x.anno, pc.use = 28:36, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 20, col.use = my.cols)


PCElbowPlot(ntiss10x.anno, num.pc = 40)
```


```{r, fig.width=10, fig.height=10}
n.pcs = 31
ntiss10x.anno <- ntiss10x.anno %>%
    RunTSNE(dims.use = 1:n.pcs, seed.use = 10, perplexity=30)

TSNEPlot(object = ntiss10x.anno, do.label = T, pt.size = 1.2, label.size = 4, group.by = 'free_annotation', no.legend = TRUE)

```

```{r, fig.width=10, fig.height=10}
# Batch and sample effects
TSNEPlot(object = ntiss10x.anno, do.return = TRUE, group.by = "channel", no.legend = T)
TSNEPlot(object = ntiss10x.anno, do.return = TRUE, group.by = "region")
TSNEPlot(object = ntiss10x.anno, do.return = TRUE, group.by = "tissue")
TSNEPlot(object = ntiss10x.anno, do.return = TRUE, group.by = "area")
TSNEPlot(object = ntiss10x.anno, do.return = TRUE, group.by = "compartment")

```


```{r, fig.width=10, fig.height=10}

genes_to_check = toupper(c('Epcam', 'Pecam1', 'Ptprc', 'Col1a1'))

FeaturePlot(ntiss10x.anno, genes_to_check, pt.size = 1, nCol = 2, cols.use = my.cols)

```

##Reassign clusters based on cell type


```{r, fig.width=10, fig.height=10}

clusters = 1:length(table(ntiss10x.anno@meta.data[,'free_annotation']))
names(clusters) = names(sort(table(ntiss10x.anno@meta.data[,'free_annotation']), decreasing = T))
ntiss10x.anno@meta.data[,'cluster_ids'] <- clusters[ntiss10x.anno@meta.data[,'free_annotation']]
clusters <- clusters[ntiss10x.anno@meta.data[,'free_annotation']]
names(clusters) <- rownames(ntiss10x.anno@meta.data)
ntiss10x.anno@ident <- as.factor(clusters)


```


```{r, fig.width=10, fig.height=10}
TSNEPlot(object = ntiss10x.anno, do.label = T, pt.size = 1.2, label.size = 4, no.legend = TRUE)

```


# Save the Robject for later
When you save the annotated tissue, please give it a name.

```{r, fig.width=10, fig.height=10}
filename = paste0("seurat/droplet_normal_lung_seurat_ntiss10x.P2.anno.20190129.RC2.Robj")
print(filename)
ntiss10x.P2.anno <- ntiss10x.anno
save(ntiss10x.P2.anno, file=filename)
```

```{r, fig.width=10, fig.height=10}
# To reload a saved object
# load(file=filename)
```
